File: epoxy-processor/src/main/java/com/airbnb/epoxy/LithoSpecProcessor.java
package com.airbnb.epoxy;

import com.airbnb.epoxy.GeneratedModelWriter.BeforeBuildCallback;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeSpec.Builder;

import java.lang.annotation.Annotation;
import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;

import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;

import static com.airbnb.epoxy.ClassNames.EPOXY_LITHO_MODEL;
import static com.airbnb.epoxy.Utils.getAnnotationClass;

class LithoSpecProcessor {

  private final Elements elementUtils;
  private final Types typeUtils;
  private final ConfigManager configManager;
  private final ErrorLogger errorLogger;
  private final GeneratedModelWriter modelWriter;
  private Class<? extends Annotation> layoutSpecAnnotationClass;

  LithoSpecProcessor(Elements elementUtils, Types typeUtils,
      ConfigManager configManager, ErrorLogger errorLogger, GeneratedModelWriter modelWriter) {

    this.elementUtils = elementUtils;
    this.typeUtils = typeUtils;
    this.configManager = configManager;
    this.errorLogger = errorLogger;
    this.modelWriter = modelWriter;
  }

  Collection<LithoModelInfo> processSpecs(RoundEnvironment roundEnv) {
    Map<TypeElement, LithoModelInfo> modelInfoMap = new LinkedHashMap<>();

    if (!hasLithoEpoxyDependency()) {
      // If the epoxy-litho module has not been included then we don't have access to the Epoxy
      // litho model and can't build a model that extends it
      return new ArrayList<>();
    }

    layoutSpecAnnotationClass =
        getAnnotationClass(ClassNames.LITHO_ANNOTATION_LAYOUT_SPEC);
    if (layoutSpecAnnotationClass == null) {
      // There is no dependency on Litho so there aren't any litho components to check for
      return new ArrayList<>();
    }

    for (Element lithoLayout : roundEnv.getElementsAnnotatedWith(layoutSpecAnnotationClass)) {
      if (!(lithoLayout instanceof TypeElement)) {
        continue;
      }

      TypeElement typeElement = (TypeElement) lithoLayout;
      modelInfoMap.put(typeElement, new LithoModelInfo(typeUtils, elementUtils, typeElement));
    }

    Class<? extends Annotation> propClass = getAnnotationClass(ClassNames.LITHO_ANNOTATION_PROP);
    for (Element propElement : roundEnv.getElementsAnnotatedWith(propClass)) {
      LithoModelInfo lithoModelInfo = getModelInfoForProp(modelInfoMap, propElement);
      if (lithoModelInfo != null) {
        lithoModelInfo.addProp(propElement);
      }
    }

    for (Entry<TypeElement, LithoModelInfo> modelInfoEntry : modelInfoMap.entrySet()) {
      try {
        final LithoModelInfo modelInfo = modelInfoEntry.getValue();
        modelWriter.generateClassForModel(modelInfo, new BeforeBuildCallback() {
          @Override
          public void modifyBuilder(Builder builder) {
            updateGeneratedClassForLithoComponent(modelInfo, builder);
          }
        });
      } catch (Exception e) {
        errorLogger.logError(e, "Error generating model classes");
      }
    }

    return modelInfoMap.values();
  }

  private boolean hasLithoEpoxyDependency() {
    // Only true if the epoxy-litho module is included in dependencies
    return Utils.getClass(EPOXY_LITHO_MODEL) != null;
  }

  private void updateGeneratedClassForLithoComponent(LithoModelInfo modelInfo,
      Builder classBuilder) {
    // Adding the "buildComponent" method
    MethodSpec.Builder methodBuilder = MethodSpec.methodBuilder("buildComponent")
        .addAnnotation(Override.class)
        .addModifiers(Modifier.PROTECTED)
        .returns(
            ParameterizedTypeName.get(ClassNames.LITHO_COMPONENT, modelInfo.lithoComponentName)
        )
        .addParameter(ClassNames.LITHO_COMPONENT_CONTEXT, "context")
        .addCode("return $T.create(context)", modelInfo.lithoComponentName);

    for (AttributeInfo attributeInfo : modelInfo.attributeInfo) {
      methodBuilder.addCode(".$L($L)", attributeInfo.getName(), attributeInfo.getName());
    }

    methodBuilder.addStatement(".build()");

    classBuilder.addMethod(methodBuilder.build());
  }

  private LithoModelInfo getModelInfoForProp(Map<TypeElement, LithoModelInfo> modelInfoMap,
      Element propElement) {
    Element methodElement = propElement.getEnclosingElement();
    if (methodElement == null) {
      return null;
    }

    Element classElement = methodElement.getEnclosingElement();

    if (classElement.getAnnotation(layoutSpecAnnotationClass) == null) {
      return null;
    }

    return modelInfoMap.get(classElement);
  }
}

File: epoxy-processor/src/main/java/com/airbnb/epoxy/LithoSpecProcessor.kt
package com.airbnb.epoxy

import com.airbnb.epoxy.ClassNames.EPOXY_LITHO_MODEL
import com.airbnb.epoxy.GeneratedModelWriter.BuilderHooks
import com.airbnb.epoxy.Utils.getAnnotationClass
import com.squareup.javapoet.MethodSpec
import com.squareup.javapoet.ParameterizedTypeName
import com.squareup.javapoet.TypeSpec.Builder
import java.util.*
import javax.annotation.processing.RoundEnvironment
import javax.lang.model.element.Element
import javax.lang.model.element.Modifier
import javax.lang.model.element.TypeElement
import javax.lang.model.util.Elements
import javax.lang.model.util.Types

internal class LithoSpecProcessor(
    private val elementUtils: Elements,
    private val typeUtils: Types,
    private val errorLogger: ErrorLogger,
    private val modelWriter: GeneratedModelWriter
) {
    private var layoutSpecAnnotationClass: Class<out Annotation>? = null

    fun processSpecs(roundEnv: RoundEnvironment): Collection<LithoModelInfo> {

        if (!hasLithoEpoxyDependency()) {
            // If the epoxy-litho module has not been included then we don't have access to the Epoxy
            // litho model and can't build a model that extends it
            return emptyList()
        }

        layoutSpecAnnotationClass = getAnnotationClass(ClassNames.LITHO_ANNOTATION_LAYOUT_SPEC)
        if (layoutSpecAnnotationClass == null) {
            // There is no dependency on Litho so there aren't any litho components to check for
            return emptyList()
        }

        val modelInfoMap = LinkedHashMap<TypeElement, LithoModelInfo>()
        roundEnv.getElementsAnnotatedWith(layoutSpecAnnotationClass)
            .filterIsInstance<TypeElement>()
            .forEach { modelInfoMap.put(it, LithoModelInfo(typeUtils, elementUtils, it)) }

        val propClass = getAnnotationClass(ClassNames.LITHO_ANNOTATION_PROP)
        val hashCodeValidator = HashCodeValidator(typeUtils, elementUtils)
        for (propElement in roundEnv.getElementsAnnotatedWith(propClass)) {
            val lithoModelInfo = getModelInfoForProp(modelInfoMap, propElement)
            lithoModelInfo?.addProp(propElement, hashCodeValidator)
        }

        for ((_, modelInfo) in modelInfoMap) {
            try {
                modelWriter.generateClassForModel(modelInfo, object : BuilderHooks() {
                    override fun beforeFinalBuild(builder: Builder) {
                        updateGeneratedClassForLithoComponent(modelInfo, builder)
                    }
                })
            } catch (e: Exception) {
                errorLogger.logError(e, "Error generating model classes")
            }

        }

        return modelInfoMap.values
    }

    // Only true if the epoxy-litho module is included in dependencies
    private fun hasLithoEpoxyDependency(): Boolean =
        getTypeMirror(EPOXY_LITHO_MODEL.reflectionName(), elementUtils) != null

    private fun updateGeneratedClassForLithoComponent(
        modelInfo: LithoModelInfo,
        classBuilder: Builder
    ) {
        // Adding the "buildComponent" method
        val methodBuilder = MethodSpec.methodBuilder("buildComponent")
            .addAnnotation(Override::class.java)
            .addModifiers(Modifier.PROTECTED)
            .returns(
                ParameterizedTypeName.get(
                    ClassNames.LITHO_COMPONENT,
                    modelInfo.lithoComponentName
                )
            )
            .addParameter(ClassNames.LITHO_COMPONENT_CONTEXT, "context")
            .addCode("return \$T.create(context)", modelInfo.lithoComponentName)

        for (attributeInfo in modelInfo.attributeInfo) {
            methodBuilder.addCode(
                ".\$L(\$L)", attributeInfo.fieldName,
                attributeInfo.fieldName
            )
        }

        methodBuilder.addStatement(".build()")

        classBuilder.addMethod(methodBuilder.build())
    }

    private fun getModelInfoForProp(
        modelInfoMap: Map<TypeElement, LithoModelInfo>,
        propElement: Element
    ): LithoModelInfo? {
        val methodElement = propElement.enclosingElement ?: return null

        val classElement = methodElement.enclosingElement

        return if (classElement.getAnnotation(layoutSpecAnnotationClass) == null) {
            null
        } else {
            modelInfoMap[classElement]
        }

    }
}
